Add handle for rpmbuild option

peterbaouoft opened the pull request:
Adds an option with rpmbuild. Also added some unit tests to it. More detailed description are in the commit msgs. Wanted to have this submitted first for some early suggestions =).

Things left to do:
- integration test to prove the functionality (need help). My current way of testing is... by manually running rpmdistro-gitoverlay build for origin package .... which takes very long time, Any suggestions for faster testing? =) Thanks
- fully verify the functionality of --rpmbuild-opts  

**Note**: I added unit tests cuz.. I was not able to find existing ones yet.. but if I missed the existing test-suite somehow, please let me know. (The tests were similar idea as atomic tests anyway, so it was not too much of work =p)

**Also**: I am still pretty new to this repo, so if I happened to do something very wrong, please let me know as well =). Thanks.

Issue: #44

cgwalters commented on doc/example-overlay.yml:
```diff
@@ -30,6 +30,14 @@ components:
+    # into rpmbuild. E.g the example below is equivalent to
+    # --define make_redistributable 0
+    defines:
+        make_redistributable: 0
```
This is relatively minor but I'd quote the `0` as `"0"` - in general working with YAML you need to think about what values are strings vs numbers.

cgwalters replied:
I think specifically I'd like to throw an error if the provided value isn't already a string, if we don't do that already.  Can you just make that change (or verify it correctly fails today)?

peterbaouoft replied:
Ha, that was handled in the unit test: 
https://github.com/cgwalters/rpmdistro-gitoverlay/pull/45/files#diff-4ef342fb5600e129983542219b0ee727R44. And yup, I think erroring out will be a better option too =).

cgwalters commented on rdgo/utils.py:
```diff
@@ -33,6 +33,24 @@ def log(msg):
     sys.stdout.write('\n')
     sys.stdout.flush()
 
+def convert_key_pair_into_commands(key_value_pairs, command_name):
```
I'm fine keeping it this way, it feels a bit overly generic though given nothing else uses it.

peterbaouoft replied:
Yup, I would also agree. I was practising this generic style of coding recently, and really wanted to show it to see how you think about it =).  Will do the change too.

cgwalters commented on tests/unit/test_buildoptions.py:
```diff
@@ -0,0 +1,53 @@
+
+no_mock = True
+try:
+    from unittest.mock import ANY, patch, call
```
Definitely confusing that we're using the RPM mock and using the "mock" unit test framework 😉

cgwalters replied:
That said is there a reason to support not having the mock unit test package?  I'm fine making it a hard dependency for running the tests.

Did you copy/adapt this unit code scaffolding from somewhere else that was doing this?

peterbaouoft commented:
Thanks @cgwalters for review =). Will address those comments shortly(after finishing the intern presentation preparation :p). One aside question, any suggestions for how to add an integration test to this feature?

 I am not sure if directly verifying pulling an actual package with overlay.yml will be worth it. Would like to know an easier/minimal way too =)

cgwalters commented:
It'd be nice to have integration tests...but let's not block right now on creating them for this.

peterbaouoft commented:
> but let's not block right now on creating them for this.

SGTM, thanks for the comments. Force pushed new commits addressing comments above. Also tested with openshift origin package:

With no rpmbuildopts, it will not have `OS_ONLY_BUILD_PLATFORMS`
```
Executing: /usr/bin/mock --configdir /tmp/tmp3s9bk7n4mockchain/configs/centos-7-and-extras-x86_64 --uniqueext mockchain-2834 -r /tmp/tmp3s9bk7n4mockchain/configs/centos-7-and-extras-x86_64/centos-7-and-extras-x8
6_64.cfg --nocheck --old-chroot --resultdir /rdgo/build-0/origin-3.10.0.alpha.0.1120-bcac7f391b820a4514002c377d2cb47166909951 --no-cleanup-after --rpmbuild-opts "--define \"make_redistributable 0\"" /rdgo/build-
0/origin-3.10.0.alpha.0.1120-bcac7f391b820a4514002c377d2cb47166909951/srpm/origin-3.10.0.alpha.0.1120-bcac7f391b820a4514002c377d2cb47166909951.el7.centos.src.rpm    
.................................
+ BUILD_PLATFORM=linux/amd64
+ OS_ONLY_BUILD_PLATFORMS=linux/amd64
+ OS_GIT_VERSION=
+ OS_GIT_COMMIT=
+ OS_GIT_MAJOR=
+ OS_GIT_MINOR=
+ OS_GIT_TREE_STATE=
+ OS_BUILD_RELEASE_ARCHIVES=n
+ make build-cross
```

peterbaouoft commented:
gentle pin. :) @cgwalters

cgwalters commented on rdgo/utils.py:
```diff
@@ -33,6 +33,26 @@ def log(msg):
+
+        # Convert the key value pair into '--define key value"' command format
+        # e.g a key, value: foo bar, will be evaluated into '--define "foo bar"'
+        command = "--define" + " " + '\"' + key + " " + value + '\"'
```
How about: `command = '--define "{} {}"'.format(key, value)` ?

cgwalters commented on rdgo/mockchain.py:
```diff
@@ -251,6 +251,12 @@ def do_one_build(self, pkg):
                         '--no-cleanup-after'])
+
+        if pkg.rpmbuildopts:
+            # Collect all rpmbuildopts and join them into one string
```
Hmm.  This is really a `mock` problem but this is opening us up to painful quoting issues if the defines ever have any `"` characters etc.  Nothing to worry about now though.

(To fix it we'd have to change mock to have e.g. `'--rpmbuild-opt=--define "foo bar"' '--rpmbuild-opt=--define "blah baz"' or so)

cgwalters commented:
Just one thing to tweak, otherwise LGTM!   Sorry about missing your reply earlier.

peterbaouoft commented:
Pushed fixup above to address comments, let me know if squashing is needed =)

cgwalters commented:
This repo doesn't have homu set up - mind squashing the fixups?

peterbaouoft commented:
> This repo doesn't have homu set up - mind squashing the fixups?

Sure, done above =).

cgwalters commented:
Thanks!

peterbaouoft commented:
> Thanks!

Thanks for the review =)
